Mathlib/AlgebraicTopology/SimplicialSet/Dimension.lean

```diff
@@ -0,0 +1,131 @@
+
+section
+
+variable (X : SSet.{u}) (d : ℕ) [X.HasDimensionLT d] (n : ℕ)
+
+lemma degenerate_eq_top_of_hasDimensionLT (hn : d ≤ n) : X.degenerate n = ⊤ :=
+  HasDimensionLT.degenerate_eq_top n hn
+
+lemma nonDegenerate_eq_bot_of_hasDimensionLT (hn : d ≤ n) : X.nonDegenerate n = ⊥ := by
+  simp [nonDegenerate, X.degenerate_eq_top_of_hasDimensionLT d n hn]
+
+lemma dim_lt_of_nonDegenerate {n : ℕ} (x : X.nonDegenerate n) (d : ℕ)
+    [X.HasDimensionLT d] : n < d := by
+  by_contra!
+  obtain ⟨x, hx⟩ := x
+  simp [X.nonDegenerate_eq_bot_of_hasDimensionLT d n this] at hx
+
+lemma dim_le_of_nonDegenerate {n : ℕ} (x : X.nonDegenerate n) (d : ℕ)
+    [X.HasDimensionLE d] : n ≤ d :=
+  Nat.le_of_lt_succ (X.dim_lt_of_nonDegenerate x (d + 1))
+
+lemma hasDimensionLT_of_le (hn : d ≤ n := by cutsat) : HasDimensionLT X n where
+  degenerate_eq_top i hi :=
+    X.degenerate_eq_top_of_hasDimensionLT d i (hn.trans hi)
+
+instance [HasDimensionLT X n] : HasDimensionLT X (n + 1) :=
+  X.hasDimensionLT_of_le n _
+
+end
+
+namespace Subcomplex
+
+variable {X : SSet.{u}}
+
+instance (d : ℕ) [X.HasDimensionLT d] (A : X.Subcomplex) : HasDimensionLT A d where
+  degenerate_eq_top (n : ℕ) (hd : d ≤ n) := by
+    ext x
+    simp [A.mem_degenerate_iff, X.degenerate_eq_top_of_hasDimensionLT d n hd]
+
+lemma le_iff_of_hasDimensionLT (A B : X.Subcomplex) (d : ℕ) [X.HasDimensionLT d] :
+    A ≤ B ↔ ∀ (i : ℕ) (_ : i < d), A.obj _ ∩ X.nonDegenerate i ⊆ B.obj (op ⦋i⦌) := by
+  refine ⟨fun h i hi a ⟨ha, _⟩ ↦ h _ ha, fun h ↦ ?_⟩
+  rw [le_iff_contains_nonDegenerate]
+  rintro n x hx
+  exact h _ (X.dim_lt_of_nonDegenerate x d) ⟨hx, x.prop⟩
+
+lemma eq_top_iff_of_hasDimensionLT (A : X.Subcomplex) (d : ℕ) [X.HasDimensionLT d] :
+    A = ⊤ ↔ ∀ (i : ℕ) (_ : i < d), X.nonDegenerate i ⊆ A.obj _ := by
+  simp [← top_le_iff, le_iff_of_hasDimensionLT ⊤ A d]
+
+end Subcomplex
+
+lemma hasDimensionLT_of_mono {X Y : SSet.{u}} (f : X ⟶ Y) [Mono f] (d : ℕ)
+    [Y.HasDimensionLT d] : X.HasDimensionLT d where
+  degenerate_eq_top n hn := by
+    ext x
+    rw [← degenerate_iff_of_isIso (Subcomplex.toRange f),
+      Subcomplex.mem_degenerate_iff, Y.degenerate_eq_top_of_hasDimensionLT d n hn]
+    simp
+
+lemma Subcomplex.hasDimensionLT_of_le
+    {X : SSet.{u}} {A B : X.Subcomplex} (h : A ≤ B) (d : ℕ) [HasDimensionLT B d] :
+    HasDimensionLT A d :=
+  hasDimensionLT_of_mono (Subcomplex.homOfLE h) d
+
+lemma hasDimensionLT_of_epi {X Y : SSet.{u}} (f : X ⟶ Y) [Epi f] (d : ℕ)
+    [X.HasDimensionLT d] : Y.HasDimensionLT d where
+  degenerate_eq_top n hn := by
+    ext y
+    simp only [Set.top_eq_univ, Set.mem_univ, iff_true]
+    obtain ⟨x, rfl⟩ := epi_iff_surjective (f := (f.app (op ⦋n⦌))).1 inferInstance y
+    apply degenerate_app_apply
+    simp [X.degenerate_eq_top_of_hasDimensionLT d n hn]
+
+lemma hasDimensionLT_iff_of_iso {X Y : SSet.{u}} (e : X ≅ Y) (d : ℕ) :
+    X.HasDimensionLT d ↔ Y.HasDimensionLT d :=
+  ⟨fun _ ↦ hasDimensionLT_of_epi e.hom d, fun _ ↦ hasDimensionLT_of_epi e.inv d⟩
```

Comment:
maybe this version is slightly more practical?

```suggestion
lemma hasDimensionLT_iff_of_isIso {X Y : SSet.{u}} (e : X ⟶ Y) [IsIso e] (d : ℕ) :
    X.HasDimensionLT d ↔ Y.HasDimensionLT d :=
  ⟨fun _ ↦ hasDimensionLT_of_epi e d, fun _ ↦ hasDimensionLT_of_epi (inv e) d⟩
```

Reply:
Thanks for the reviews!

I would prefer keep the `iso` version, as I believe the most common phrasing in mathlib for such properties is in terms of `iso` rather than `isIso`.